src/Agent/NewRelic/Agent/Extensions/NewRelic.Agent.Extensions/Helpers/SearchRequestWrapperBase.cs

```diff
@@ -0,0 +1,274 @@
+// Copyright 2020 New Relic, Inc. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Threading.Tasks;
+using NewRelic.Agent.Api;
+using NewRelic.Agent.Api.Experimental;
+using NewRelic.Agent.Extensions.Parsing;
+using NewRelic.Agent.Extensions.Providers.Wrapper;
+using NewRelic.Agent.Extensions.SystemExtensions;
+using NewRelic.Reflection;
+
+namespace NewRelic.Agent.Extensions.Helpers
+{
+    public abstract class SearchRequestWrapperBase
+    {
+        private Func<object, object> _apiCallDetailsGetter;
+        private Func<object, bool> _successGetter;
+        private Func<object, object> _exceptionGetter;
+        private Func<object, Uri> _uriGetter;
+
+        protected ConcurrentDictionary<Type, Func<object, object>> GetRequestResponseFromGeneric = new ConcurrentDictionary<Type, Func<object, object>>();
+
+        public abstract DatastoreVendor Vendor { get; } 
+
+        protected ISegment BuildSegment(int requestParamsIndex, InstrumentedMethodCall instrumentedMethodCall, ITransaction transaction)
+        {
+            var path = (string)instrumentedMethodCall.MethodCall.MethodArguments[1];
+            var request = instrumentedMethodCall.MethodCall.MethodArguments[0].ToString();
+            var requestParams = instrumentedMethodCall.MethodCall.MethodArguments[requestParamsIndex];
+            var splitPath = path.Trim('/').Split('/');
+
+            var operation = (requestParams == null) ? GetOperationFromPath(request, splitPath) : GetOperationFromRequestParams(requestParams);
+
+            var model = splitPath[0]; // For SQL datastores, "model" is the table name. For OpenSearch it's the index name.  This is often the first component of the request path, but not always.
+            if ((model.Length == 0) || (model[0] == '_')) // Per OpenSearch docs, index names aren't allowed to start with an underscore, and the first component of the path can be an operation name in some cases, e.g. "_bulk" or "_msearch"
```

Comment:
What if it's a sql table named `_some_stupid_name` ?

Reply:
Based on the docs for Elastic/OpenSearch they won't start with an underscore.